include/picongpu/plugins/openPMD/openPMDWriter.hpp

```diff
@@ -0,0 +1,1417 @@
+            const uint32_t numSlides =
+                MovingWindow::getInstance().getSlideCounter(
+                    params->currentStep );
+            globalSlideOffset.y() += numSlides * localDomain.size.y();
+
+            // globalDimensions is {x, y, z} but fields are F[z][y][x]
+            std::vector< float_64 > gridGlobalOffset( simDim, 0.0 );
+            for( uint32_t d = 0; d < simDim; ++d )
+                gridGlobalOffset.at( simDim - 1 - d ) = float_64(
+                                                            cellSize[ d ] ) *
+                    float_64( params->window.globalDimensions.offset[ d ] +
+                              globalSlideOffset[ d ] );
+
+            mesh.setGridGlobalOffset( std::move( gridGlobalOffset ) );
+            mesh.setGridUnitSI( UNIT_LENGTH );
+            mesh.setAttribute( "fieldSmoothing", "none" );
+        }
+
+        template< typename ComponentType >
+        static void
+        writeField(
+            ThreadParams * params,
+            const uint32_t sizePtrType,
+            ::openPMD::Datatype openPMDType,
+            const uint32_t nComponents,
+            const std::string name,
+            void * ptr,
+            std::vector< float_64 > unit,
+            std::vector< float_64 > unitDimension,
+            std::vector< std::vector< float_X > > inCellPosition,
+            float_X timeOffset )
+        {
+            const std::string name_lookup_tpl[] = { "x", "y", "z", "w" };
+
+            /* parameter checking */
+            PMACC_ASSERT( unit.size() == nComponents );
+            PMACC_ASSERT( inCellPosition.size() == nComponents );
+            for( uint32_t n = 0; n < nComponents; ++n )
+                PMACC_ASSERT( inCellPosition.at( n ).size() == simDim );
+            PMACC_ASSERT(
+                unitDimension.size() == 7 ); // seven openPMD base units
+
+            log< picLog::INPUT_OUTPUT >( "openPMD: write field: %1% %2% %3%" ) %
+                name % nComponents % ptr;
+
+            const bool fieldTypeCorrect(
+                boost::is_same< ComponentType, float_X >::value );
+            PMACC_CASSERT_MSG(
+                Precision_mismatch_in_Field_Components__ADIOS,
+                fieldTypeCorrect );
+
+            ::openPMD::Iteration iteration =
+                params->openPMDSeries->iterations[ params->currentStep ];
+            ::openPMD::Mesh mesh = iteration.meshes[ name ];
+
+            // set mesh attributes
+            writeFieldAttributes( params, unitDimension, timeOffset, mesh );
+
+            /* data to describe source buffer */
+            GridLayout< simDim > field_layout = params->gridLayout;
+            DataSpace< simDim > field_full = field_layout.getDataSpace();
+            DataSpace< simDim > field_no_guard =
+                params->window.localDimensions.size;
+            DataSpace< simDim > field_guard =
+                field_layout.getGuard() + params->localWindowToDomainOffset;
```

Comment:
It looks this part is hard-coded for the moving window and field size == domain size. As I've just pointed out in the [discussion](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2966#issuecomment-588851575) this probably needs a fix.

Reply:
For the original ADIOS plugin, these two commits provided a fix: b6bb6cf5c37421ffeae9ecd5a176825b2d4790ce , b797c0a3cf4c22d5e3aa2512d97deb8d0ed7fe93

Reply:
The latest commit ports [b797c0a](https://github.com/ComputationalRadiationPhysics/picongpu/commit/b797c0a3cf4c22d5e3aa2512d97deb8d0ed7fe93) to the openPMD plugin, I also took the chance to use a `std::vector< float_X >` instead of a `shared_ptr` for `fieldBuffer` as you suggested some while ago. (still needs testing)
Side note: Doesn't [this line](https://github.com/ComputationalRadiationPhysics/picongpu/blob/b797c0a3cf4c22d5e3aa2512d97deb8d0ed7fe93/include/picongpu/plugins/adios/ADIOSWriter.hpp#L1256) leak the memory previously held by `dstBuffer`?

Reply:
Thanks for porting. No, it does not leak memory since `dstBuffer` does not own memory. For "normal" domain-size fields `dstBuffer` just reuses memory of `params->fieldBfr`, this latter pointer is never changed and later used to free up memory. For the PML fields, `nonDomainBoundStorage` owns memory and so cleans it up. Thus, in both cases `dstBuffer` is a non-owning pointer and must not be used to clean the memory. But this kind of confusion is a rather good example of why raw pointers should preferably never own memory, with smart pointers or vectors doing it instead.

Reply:
Ah alright, I missed that you weren't working on `ThreadParams::fieldBfr` directly. Thanks for checking!

Reply:
@sbastrakov is this issue solved? or does it require a fix in this PR.

Reply:
@sbastrakov PING

Reply:
For me it looks like this issue is solved.